### PR TITLE
Alternate implementation for waitForConnectedState

### PR DESCRIPTION
**Pull request by ChumpChief**
Splitting this from #14265 since it involves behavior change.  If folks aren't comfortable with this behavior change, then I'll instead close this PR and just hide the API in #14265.

Currently, waitForConnectedState mixes two waiting policies:
1. Wait for a certain amount of time
2. Wait for the browser's "online" state

However, the way it mixes these two is unintuitive and likely to surprise the caller.  For instance:
* If called while offline, will resolve immediately upon the browser going online (the "minDelay" is bypassed)
* `navigator.onLine` can easily be `false` when it resolves

Looking at the call sites, side effects like these don't seem to be intended.  This change has 3 goals:
1. Make the policies explicit: Someone reading the call site should have high confidence in the actual wait time and online/offline state before/during/after.
2. Correctness: If we really intend to have a minimum delay in these places, we should adhere to it and not violate it when online/offline status changes.
3. Reduce API surface area: This API isn't meant for general public use (especially since it has these surprising policy behaviors).  Hiding it reduces noise and risk for customers who might consider using it.

Really, any heuristic about connections/network state should live in the driver (as the driver is the layer that knows whether the network should be relevant in these activities).  For instance, the local driver shouldn't care about network in any of these flows.  But at least for this change, the logic stays in the loader layer.

### CODE REVIEW

**Review by RishhiB — APPROVED** [review]

File: packages/loader/driver-utils/src/parallelRequests.ts [hunk 1]
```diff
@@ -379,6 +379,19 @@ export class Queue<T> implements IStream<T> {
 	}
 }
 
+const waitForOnline = async (): Promise<void> => {
```
Comment (vladsud) [inline packages/loader/driver-utils/src/parallelRequests.ts:382]: This looks like adds 3 copies of the same thing in the same loader bundle (I'm actually surprised only 68 bytes added, per measurements published for this PR). Can we avoid duplicating code?

File: packages/loader/container-loader/src/connectionManager.ts [hunk 2]
```diff
@@ -574,7 +586,16 @@ export class ConnectionManager implements IConnectionManager {
...
+
+				// First, wait for the amount of time we think is appropriate.
+				await new Promise<void>((resolve) => {
+					setTimeout(resolve, delayMs);
+				});
+
+				// Then, if we believe we're offline, we assume there's no point in trying until we at least think we're online.
+				// NOTE: This isn't strictly true for drivers that don't require network (e.g. local driver).  Really this logic
+				// should probably live in the driver.
+				await waitForOnline();
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionManager.ts:598]: Glancing at the usage, I think the intention of the old code was - if network request failed, then retry with some polling frequency. But in case we are offline (and thus network error was caused by being offline), disregard polling frequency if we detect establishment of connection.

It's probably does not matter that much with our max retry of 8 seconds, but if it was in minutes, then new behavior might result (from end-user POV) as a regression, as we would not reconnect immediately on network coming back to life.

I agree with confusing behavior hidden in one function, but maybe the right fix here to replace waitForConnectedState(minDelay) usage with Promise.race(waitForOnline(), delay(miDelay)) - that's what callers expect, I think.

Comment (ChumpChief) [reply packages/loader/container-loader/src/connectionManager.ts:598]: One thing I was trying to avoid is violating the `error.retryAfterSeconds` - seems like regardless of network coming online/offline we should not try again sooner than the error specifies.  Maybe I'll split out the `retryDelayFromError` from the `delayMs` so we can wait the full duration the error requests, but otherwise race like you say.

File: packages/loader/driver-utils/src/parallelRequests.ts (hunk 1 above)
Comment (ChumpChief) [reply packages/loader/driver-utils/src/parallelRequests.ts:382]: I'll see what we can do.

IMO root problem is that both `ConnectionManager` and `ParallelRequests` (and `requestOps()`) should all live hidden in driver layer/package, where they would easily be able to share implementation without public exposure.  `ParallelRequests` and `requestOps()` can fairly easily get moved to a more legitimate driver-layer package (which I'd like to do soon), but not `ConnectionManager`.  So it's a struggle between layer cleanliness, back-compat, and duplicated code.

#14265 will reduce bundle size so I'm paying the debt at least, but I'll see if there's any good clever option.

File: packages/loader/container-loader/src/connectionManager.ts (hunk 2 above)
Comment (vladsud) [reply packages/loader/container-loader/src/connectionManager.ts:598]: Logically, we should never have retryAfterSeconds & being offline. That, of course, assumes that offline signal is accurate (which is not the case, but I do not if it produces false positives - it definitely produces false negatives).

So while your concern is valid, I'm not sure how much we should care.
But if we want to take care of that combo, we likely need to make more invasive changes to consuming code, and separate dealing with retryAfterSeconds vs. polling frequency (when we do not have retryAfterSeconds).
I'd personally prefer for code to be simpler :)

Comment (ChumpChief) [reply packages/loader/container-loader/src/connectionManager.ts:602]: Latest includes this improvement for ConnectionManager (no race required - no matter what every path should wait for online before retry).  Taking a look at ParallelRequests now.

File: packages/loader/driver-utils/src/parallelRequests.ts [hunk 3]
```diff
@@ -477,7 +490,15 @@ async function getSingleOpBatch(
 			}
 		}
 
-		await waitForConnectedState(delay);
```
Comment (ChumpChief) [inline packages/loader/driver-utils/src/parallelRequests.ts:480]: One thing that's a little odd with the current implementation is that even if we successfully receive part of the deltas, we wait with a scaling delay (same as if we had gotten an error).  This seems maybe unintentional -- probably if we successfully get part of the deltas we should just immediately turn around and request the rest.

Comment (ChumpChief) [reply packages/loader/driver-utils/src/parallelRequests.ts:480]: Took me a minute to realize this is not the case that part of the deltas was retrieved, but rather an empty set.  So this logic should be fine.  Latest does some cleanup to try to make this more obvious.

File: packages/loader/container-loader/src/connectionManager.ts [hunk 4]
```diff
@@ -569,12 +581,25 @@ export class ConnectionManager implements IConnectionManager {
 				lastError = origError;
 
 				const retryDelayFromError = getRetryDelayFromError(origError);
-				delayMs = retryDelayFromError ?? Math.min(delayMs * 2, MaxReconnectDelayInMs);
-
 				if (retryDelayFromError !== undefined) {
+					// If the error told us to wait, then we wait.
 					this.props.reconnectionDelayHandler(retryDelayFromError, origError);
+					await new Promise<void>((resolve) => {
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionManager.ts:587]: I think we have delay() function somewhere that does that.

File: packages/loader/container-loader/src/connectionManager.ts [hunk 5]
```diff
@@ -569,12 +581,25 @@ export class ConnectionManager implements IConnectionManager {
...
+						setTimeout(resolve, retryDelayFromError);
+					});
+				} else if (globalThis.navigator?.onLine !== false) {
```
Comment (vladsud) [inline packages/loader/container-loader/src/connectionManager.ts:590]: It feels like we can remove the check for globalThis.navigator?.onLine here and make it unconditional.
There is small change in behavior, but it feels like it does not matter, and code becomes simpler / smaller.

File: packages/loader/driver-utils/src/parallelRequests.ts [hunk 6]
```diff
@@ -449,8 +460,6 @@ async function getSingleOpBatch(
 		} catch (error) {
 			const canRetry = canRetryOnError(error);
 
-			lastSuccessTime = undefined;
```
Comment (vladsud) [inline packages/loader/driver-utils/src/parallelRequests.ts:452]: Is this change intentional? Feels like it changes semantics of this code

File: packages/loader/driver-utils/src/parallelRequests.ts (hunk 3 above)
Comment (vladsud) [reply packages/loader/driver-utils/src/parallelRequests.ts:480]: I do not remember much :) But your description rings the bell - we run into situations where clients knows that it needs ops, but they are in transition from ordering service to storage, and thus temporarily unavailable. So client will go in the loop asking storage for ops until it gets them. I think we want to keep circling without delay.

**Review by vladsud — APPROVED** [review]

File: packages/loader/driver-utils/src/parallelRequests.ts (hunk 6 above)
Comment (ChumpChief) [reply packages/loader/driver-utils/src/parallelRequests.ts:452]: Yes and it does - I couldn't come up with a reason that an error should permit more attempts on getting a non-empty response, or why it would even relate to it, so I opted to remove it (in most cases it will have no effect, in the only cases where it has an effect it will just be that we still give up after 30 seconds of trying)

File: packages/loader/container-loader/src/connectionManager.ts (hunk 5 above)
Comment (ChumpChief) [reply packages/loader/container-loader/src/connectionManager.ts:590]: That was my initial implementation but it introduces the performance regression you called out in your original comments.  I do think skipping the delay when transitioning to online makes sense to get a faster connection.

File: packages/loader/container-loader/src/connectionManager.ts (hunk 4 above)
Comment (ChumpChief) [reply packages/loader/container-loader/src/connectionManager.ts:587]: We do but it's in common-utils, which I expect I'll go rearrange relatively soon (need to figure out what its relationship should be relative to the new core-utils package).  For now I don't want to take new dependencies on it.

File: packages/loader/driver-utils/src/parallelRequests.ts (hunk 3 above)
Comment (ChumpChief) [reply packages/loader/driver-utils/src/parallelRequests.ts:480]: I thought about removing the delay, but assuming the transition from ordering service to storage takes some time, this delay helps us not blast too many requests at the storage before that happens.  Would need some research into what's a reasonable expectation for the transfer time before removing the delay.
